### PR TITLE
autotools: Fix static library for mingw

### PR DESCRIPTION
**Pull request by Biswa96**
This change defines FRIBIDI_LIB_STATIC while building static library
if fribidi present. Otherwise, libass static library imports fribidi
APIs with dllimport attribute.

### CODE REVIEW

**Comment by astiob** [comment]
Thanks! See also: #493. @TheOneric Could you remind me where that ended up at?

**Comment by astiob** [comment]
If I understand this patch correctly, it adds `FRIBIDI_LIB_STATIC` if static libass is enabled, but it takes effect even if shared libass is also enabled and affects the shared libass as well (because the object files are not compiled separately for static and shared). At that point, it seems that this has few advantages over simply defining the macro _always_ as proposed in #493. Technically shared-only libass builds would be slightly more efficient in calling FriBidi with this patch than with #493, but this efficiency difference hardly seems worth anything.

Could you perhaps test #493 with static and shared libass and confirm whether it works?

**Comment by Biswa96** [comment]
Yes, it also works. But as you said, it will be always defined which also removes the `__declspec(dllimport)` keyword from all fribidi function declarations irrespective static or shared build.

**Comment by TheOneric** [comment]
> Could you remind me where that [#493] ended up at?

The patch was never actually tested, we weren't sure if defining `FRIBIDI_LIB_STATIC` or `FRIBIDI_EXTERN` would be better and we hope `pkg-config` will adopt the `Cflags.private`-patch as a cleaner solution with more portable pc-files and avoiding the need for a patch on our side. I did not get a response from a pkg-config maintainer yet though.

**Comment by TheOneric** [comment]
The proposed patch is wrong. `-DFRIBIDI_LIB_STATIC` should depend on the nature of the used *FriBidi* library being linked to, whether or not *libass* is being built shared or static bears no relevance to my knowledge. If FriBidi's `declspec(dllimport)` were needed for libass' use case, defining `-DFRIBIDI_LIB_STATIC` as done here would prevent usage of any static libass build with a dynamic FriBidi *(which is done in parts our CI)* and also as noted by astiob, prevent shared libass built together with the static variant *(the default behaviour)* from being used with a shared FriBidi. Furthermore duplicating logic from our dependency into our own build is undesirable since prone to failure with older or future versions of the dependency.

In fact, FriBidi does make an attempt to get `-DFRIBIDI_LIB_STATIC` defined automatically, by adding `CFLAGS.private: -DFRIBIDI_LIB_STATIC` to its pc-file. The issue are `Cflags.private` being specific to the `pkgconf`-implementation and the capitalisation being non-standard relying on another `pkgconf`-specific property: case-insensitive field names. Also the compiler-flags are only used if the `--static` option is passed to pkgconf (which you should do when aiming for fully-static builds).  
If this is your issue, the preferred solution is to get the original `pkg-config` to adopt `Cflags.private` or similar *(which will with high certainty lead to OpenBSD's `pkg-config` and likely other maintained implementation to follow suit)* and get FriBidi to standardise the capitalisation of the field.  
There's a patch by me to add `Cflags.private` to the original `pkg-config`, see [pkg-config!13](https://gitlab.freedesktop.org/pkg-config/pkg-config/-/merge_requests/13). If you're already using `--static` in MSys2 you may adopt this as a distro-patch in MSys2 (if required together with [pkg-config!15](https://gitlab.freedesktop.org/pkg-config/pkg-config/-/merge_requests/15) to fix builds from git).  
Note that — assuming not defining `-DFRIBIDI_LIB_STATIC` were relevant — you'd still need to build libass twice if you want a (shared) version which can link to shared FriBidi and a (static) version which can link to static FriBidi. For libass specifically this might be avoidable, see next paragraph, but in general that will be required for any library with dependecies with an impactful `Cflags.private` field.

As astiob mentioned, the `dllimport` might not be relevant for libass. If you are interested in working around the need for either a pkg-config patch or two builds for different FriBidi versions on libass side, see [libass#493](https://github.com/libass/libass/issues/493). For one of 493's patches (eg defining `FRIBIDI_EXTERN` instead of `FRIBIDI_LIB_STATIC`) to be considered to applied, it needs to be tested on MS Windows to build and also to work once linked in a binary (e.g. with our [regression tests](https://github.com/libass/libass-tests)) for all combinations of static and shared libass and FriBidi. Ideally, but not necessarily with *both* current FriBidi and FriBidi version 1.0.9 *(before the export behaviour was changed)*.

---

> Yes, it also works. But as you said, it will be always defined which also removes the __declspec(dllimport) keyword from all fribidi function declarations irrespective static or shared build.

See https://github.com/libass/libass/issues/493

**Comment by astiob** [comment]
Hang on. For one thing, IIRC MSYS uses pkgconf nowadays, which already supports that field, for this very reason. For another, how is pkgconf supposed to know whether FriBidi is static or not? _We_ are the ones asking it for `--libs` and adding or not adding `--static`, but we don’t know which FriBidi library files are actually available, and neither does pkgconf. Surely our best choice of `--static` is/would be no different from the patch that was proposed here.

Anyway, I think #493 should just be merged.

**Comment by astiob** [comment]
Indeed, the linked MSYS [issue](https://github.com/msys2/MINGW-packages/issues/8817)/[PR](https://github.com/msys2/MINGW-packages/pull/8818) suggests a pure static build already works, so I assume `--static` is being passed and pkgconf is picking up the `Cflags.private` as expected.

**Comment by TheOneric** [comment]
> For another, how is pkgconf supposed to know whether FriBidi is static or not?

Someone wanting to build a fully static libass can(should) use eg `PKG_CONFIG="/usr/bin/pkg-config --static" ./configure` which will cause everything to use the flags appropiate for static-building.  
If only certain dependencies are desired to be used statically, it's more complicated. If `--static` is not specified but only a static library of a dependency is available it will fail if this dependecy has unique dependencies — even without a `Cflags.private` field. Common build systems and pkg-config implementations are poorly equipped for user-selectable mixed static/shared builds. *(In autoconf the program-author can specifcally select static flags by using `PKG_CHECK_MODULES_STATIC` instead of `PKG_CHECK_MODULES`)*.

**Comment by rcombs** [comment]
Shouldn't fribidi emit the `-D` in `cflags` (not just `cflags.private`) if built static(-only)?

**Comment by astiob** [comment]
> Shouldn't fribidi emit the `-D` in `cflags` (not just `cflags.private`) if built static(-only)?

Right. Probably. Maybe it does.

I guess I misunderstood Oneric’s reasoning for why the proposed patch is wrong, and now that I understand it better, I disagree with it (and I’m frustrated by the harshness of reality).

Anyway, let’s just merge #493 (which makes it work in all cases) and stop worrying about this.

---

> Someone wanting to build a fully static libass can(should) use eg `LDFLAGS="-static" PKG_CONFIG="/usr/bin/pkg-config --static" ./configure` which will cause everything to use the flags aproppiate for static-building. So for a fully static build the user/builder can/should set `--static`, not our configure script.

I guess this is an actual reason why `Cflags.private` isn’t “the solution”. Thanks again, Microsoft. You simply _can’t_ make a standalone library if it depends on anything else. In the general case, each library needs to know the static/dynamic choices in the entire application’s dependency tree. And the specific case where you _can_ is if you never export anything but functions and never use `dllimport`, which is precisely what we pretend FriBidi does in #493 (and what we do ourselves in libass).

Earlier, it sounded like it all hinged on `Cflags.private`, but now it seems like that’s not it. It already has them, they already work, but that’s not enough. And even if the `Cflags.private` weren’t there, if the user is gonna have to add custom build-system flags anyway, they might as well add `FRIBIDI_LIB_STATIC` directly.

FWIW I think normally people do expect—I would—that the binaries built by `./configure --enable-static` are suitable for a fully static build, without any extra arcane incantations. Much like how `--enable-shared` is expected to be suitable for a fully dynamic build. (As you point out, on Windows, _neither_ can be used in a mixed build in the general case.) So I would agree with using `FRIBIDI_LIB_STATIC` for `--enable-static` if it mattered; but the problem would be `--enable-static --enable-shared`.

Looking again at our Autoconf-generated `configure`, I don’t see any `--static` flags for pkg-config, so I assume my earlier conclusion that it was already used given `--disable-shared` was wrong. I guess it’s because `pkg-config --static` never affected static output libraries before pkgconf added `Cflags.private` and because there’s no obvious reason to want it except as a kludge for this `dllimport` stuff on Windows. That MSYS PR must’ve not actually succeeded in fixing their static libass/FriBidi build, which must’ve prompted this libass PR.

**Comment by TheOneric** [comment]
> Shouldn't fribidi emit the -D in cflags (not just cflags.private) if built static(-only)?

FriBidi is being built static-only; iiuc it's using the regular fribidi package from MSys2, which ships both static and shared with one pc-file.

> Anyway, let’s just merge #493 (which makes it work in all cases)

If it gets actually tested before, sure.

> the problem would be `--enable-static --enable-shared`.

This is the default, without any flags needing to get passed.

> I guess I misunderstood Oneric’s reasoning for why the proposed patch is wrong.

Let me rephrase my argument and trim the additional digressions.

If we assume both omitting and setting `-DFRIBIDI_LIB_STATIC` depending on the FriBidi-build used is important, then the patch is wrong because it will create a shared libass library which can't link to shared fribidi – by default even and because furthermore it makes it *impossible* to build a static libass which can link to a shared FriBidi *(like our CI does)* without patching our build system. In this case a least-bad solution relates to the other bits omitted this time.

If only omitting is a problem but never setting the flag (as speculated in #493), then the patch is wrong because with `--disable-static` a shared libass-library is produced which cannot link to a static FriBidi, eventhough we expect both static and shared FriBidi to work and additionally the patch is overly complex about setting `-DFRIBIDI_LIB_STATIC`. In this case #493 would be the correct solution.
